### PR TITLE
feat: add env var to bypass hooks execution

### PR DESCRIPTION
**Pull request by bhanuprasadcherukuvada**
ref -
#87

### CODE REVIEW

File: simple-git-hooks.js [hunk 1]
```diff
@@ -178,7 +178,13 @@ function _setHook(hook, command, projectRoot=process.cwd()) {
         return
     }
 
-    const hookCommand = "#!/bin/sh\n" + command
+    const prependScript =
```
Comment (toplenboren) [inline simple-git-hooks.js:181]: I think its better to export this as const, and use in tests, so we avoid code repeating

File: simple-git-hooks.test.js [hunk 2]
```diff
@@ -126,108 +127,196 @@ function getInstalledGitHooks(hooksDir) {
...
+      '    echo "[INFO] SIMPLE_GIT_HOOKS is set to 0, skipping hook."\n' +
+      "    exit 0\n" +
+      "fi\n\n"}exit 1`, 'pre-push':`${"#!/bin/sh\n\n" +
+      'if [ "$SIMPLE_GIT_HOOKS" = "0" ]; then\n' +
+      '    echo "[INFO] SIMPLE_GIT_HOOKS is set to 0, skipping hook."\n' +
+      "    exit 0\n" +
+      "fi\n\n"}exit 1`}))
 })
 
+afterEach(() => {
```
Comment (toplenboren) [inline simple-git-hooks.test.js:170]: I think we can use only one afterEach() if we use forEach loop here

**Review by toplenboren — COMMENTED** [review]
Hi! thanks for the PR! Looks good, but have a few points that can be polished :-)

File: simple-git-hooks.test.js [hunk 3]
```diff
@@ -275,7 +291,60 @@ test.each([
...
     spc.setHooksFromConfig(projectWithCustomConfigurationFilePath, args)
     const installedHooks = getInstalledGitHooks(path.normalize(path.join(projectWithCustomConfigurationFilePath, '.git', 'hooks')))
-    expect(JSON.stringify(installedHooks)).toBe(JSON.stringify({'pre-commit':`#!/bin/sh\nexit 1`, 'pre-push':`#!/bin/sh\nexit 1`}))
-
-    removeGitHooksFolder(projectWithCustomConfigurationFilePath)
+    expect(JSON.stringify(installedHooks)).toBe(JSON.stringify({'pre-commit':`${spc.PREPEND_SCRIPT}exit 1`, 'pre-push':`${spc.PREPEND_SCRIPT}exit 1`}))
 })
+
+
+test("bypasses hooks when SKIP_SIMPLE_GIT_HOOKS is set to 1", () => {
```
Comment (toplenboren) [inline simple-git-hooks.test.js:298]: You can also try grouping these test cases using `describe` syntax. It it a good practice

File: simple-git-hooks.test.js [hunk 4]
```diff
@@ -275,7 +291,60 @@ test.each([
...
+  execSync("git init \
+    && git config user.name github-actions \
+    && git config user.email github-actions@github.com", { cwd: projectWithConfigurationInPackageJsonPath });
+
+   createGitHooksFolder(projectWithConfigurationInPackageJsonPath);
+
+   spc.setHooksFromConfig(projectWithConfigurationInPackageJsonPath);
+   process.env.SKIP_SIMPLE_GIT_HOOKS = '1';  // Set environment variable
```
Comment (toplenboren) [inline simple-git-hooks.test.js:306]: I think you might need to unset variable at the end of test case.. Maybe move to after Each ? Not sure

File: simple-git-hooks.test.js [hunk 5]
```diff
@@ -126,108 +130,122 @@ function getInstalledGitHooks(hooksDir) {
...
+
 test('creates git hooks if configuration is correct from simple-git-hooks.cjs', () => {
     createGitHooksFolder(projectWithConfigurationInSeparateCjsPath)
 
     spc.setHooksFromConfig(projectWithConfigurationInSeparateCjsPath)
     const installedHooks = getInstalledGitHooks(path.normalize(path.join(projectWithConfigurationInSeparateCjsPath, '.git', 'hooks')))
-    expect(JSON.stringify(installedHooks)).toBe(JSON.stringify({'pre-commit':`#!/bin/sh\nexit 1`, 'pre-push':`#!/bin/sh\nexit 1`}))
-
-    removeGitHooksFolder(projectWithConfigurationInSeparateCjsPath)
+    expect(isEqual(installedHooks, { 'pre-commit': TEST_SCRIPT, 'pre-push': TEST_SCRIPT })).toBe(true);
```
Comment (toplenboren) [inline simple-git-hooks.test.js:176]: I think we can extract `{ 'pre-commit': TEST_SCRIPT, 'pre-push': TEST_SCRIPT }` as variable too

File: simple-git-hooks.test.js [hunk 6]
```diff
@@ -275,7 +292,77 @@ test.each([
...
     spc.setHooksFromConfig(projectWithCustomConfigurationFilePath, args)
     const installedHooks = getInstalledGitHooks(path.normalize(path.join(projectWithCustomConfigurationFilePath, '.git', 'hooks')))
-    expect(JSON.stringify(installedHooks)).toBe(JSON.stringify({'pre-commit':`#!/bin/sh\nexit 1`, 'pre-push':`#!/bin/sh\nexit 1`}))
-
-    removeGitHooksFolder(projectWithCustomConfigurationFilePath)
+    expect(JSON.stringify(installedHooks)).toBe(JSON.stringify({ 'pre-commit': TEST_SCRIPT, 'pre-push': TEST_SCRIPT }))
+    expect(isEqual(installedHooks, { 'pre-commit': TEST_SCRIPT, 'pre-push': TEST_SCRIPT })).toBe(true);
 })
+
+describe('Tests for skipping git hooks using SKIP_SIMPLE_GIT_HOOKS env var', () => {
```
Comment (toplenboren) [inline simple-git-hooks.test.js:299]: It's nice to see you taking a look to jest `describe` syntax here! 

But it makes no sense if it is used only in your tests. Let me try to be more clear on what I meant in the previous review:

So, we have our test code setup like this:

```(js)
// Get project root directory

test('getProjectRootDirectory returns correct dir in typical case:', () => {
    expect(spc.getProjectRootDirectoryFromNodeModules('var/my-project/node_modules/simple-git-hooks')).toBe('var/my-project')
})

test('getProjectRootDirectory returns correct dir when used with windows delimiters:', () => {
    expect(spc.getProjectRootDirectoryFromNodeModules('user\\allProjects\\project\\node_modules\\simple-git-hooks')).toBe('user/allProjects/project')
})
```

We can use describe to make it look like this:

```
describe('Get Project Root Directory tests', () => {
  it('returns correct dir in typical case:', () => {
    expect(spc.getProjectRootDirectoryFromNodeModules('var/my-project/node_modules/simple-git-hooks')).toBe('var/my-project')
})

it('returns correct dir when used with windows delimiters:', () => {
    expect(spc.getProjectRootDirectoryFromNodeModules('user\\allProjects\\project\\node_modules\\simple-git-hooks')).toBe('user/allProjects/project')
})
})
```

This way tests can be grouped, which allows us to see pretty output: here is an example if we do the manipulations above

Screenshot 2023-10-07 at 21 32 15

--

I propose two solutions:
- refactor all the tests to use describe syntax
- or simply leave your tests as is (do not use describe syntax) - I will do refactoring later :-)

File: simple-git-hooks.test.js [hunk 7]
```diff
@@ -275,7 +292,77 @@ test.each([
...
+
+    const GIT_USER_NAME = "github-actions";
```
Comment (toplenboren) [inline simple-git-hooks.test.js:301]: I think we can do this in GitHub actions itself, do not see why we need some GH actions specific code in repo

**Review by toplenboren — APPROVED** [review]

**Comment by toplenboren** [comment]
> @toplenboren any expected timeline for merging this PR ?

Yep. gonna do it this week

I think you can merge it yourself, I'll just test it and do the release with your feature

**Comment by jochemkeller** [comment]
Hi @toplenboren, thanks for merging the PR and many thanks to @bhanuprasadcherukuvada for adding the feature 🙏 it's very handy for my current use-case!

Can I kindly request that this feature be made available through a new NPM version? After some debugging I found out that this feature isn't available in the current latest version of simple-git-hooks, even though the GitHub `master` README documentation shows that the feature should be available 🤪 that was a confusing 15 minutes there..

File: package.json [hunk 8]
```diff
@@ -44,6 +44,7 @@
     "clean-publish": "^4.2.0",
     "eslint": "^7.19.0",
     "jest": "^26.6.3",
-    "lint-staged": "^10.5.4"
+    "lint-staged": "^10.5.4",
+    "lodash.isequal": "^4.5.0"
```
Comment (JounQin) [inline package.json:48]: What is this? I saw no usage?

**Pull request merged**